### PR TITLE
Fix compilation of GpuTopKOp

### PR DESCRIPTION
**Pull request by abergeron**
Partial fix for #6511 

However, I also found out that GpuTopKOp doesn't have any test.  This should be rectified before we declare this fixed since there might be other problems in the code than just surface compilation errors.

@khaotik You were the one that wrote this, can you also make some tests please?

### CODE REVIEW

**Comment by f0k** [comment]
You'll probably also need to change `yi` to `(yi, yi)` in https://github.com/Theano/Theano/pull/6513/files#diff-9642552a6be449fc3caf31341dc1d8d7R190.

**Comment by nouiz** [comment]
This isn't enough to fix all cases. I'm adding some tests and I have some
failures. Try it at your own risk for now.

**Comment by nouiz** [comment]
I disabled GpuTopK in the last commit. I would merge this PR when CI pass and let the fix in another PR.

**Comment by abergeron** [comment]
Flake8 is not happy:
```
/home/travis/build/Theano/Theano/theano/gpuarray/sort.py:11:1: F401 'register_opt' imported but unused
/home/travis/build/Theano/Theano/theano/gpuarray/sort.py:11:1: F401 'register_opt2' imported but unused
/home/travis/build/Theano/Theano/theano/gpuarray/sort.py:303:1: E265 block comment should start with '# '
/home/travis/build/Theano/Theano/theano/gpuarray/sort.py:305:1: E265 block comment should start with '# '
/home/travis/build/Theano/Theano/theano/gpuarray/tests/test_sort.py:5:1: F401 'test_ctx_name' imported but unused
```

File: theano/gpuarray/sort.py [hunk 1]
```diff
@@ -120,7 +135,8 @@ def build_kernel(fname, kname, subs):
             ) as f:
                 kernel_src = f.read()
             ker = Kernel(
-                code=Template(common_src + kernel_src).substitute(**subs),
+                code=("#include <cluda.h>\n" +
```
Comment (khaotik) [inline theano/gpuarray/sort.py]: could be cleaner putting this inside `theano/gpuarray/c_code/topk_common.cuh`

File: theano/gpuarray/sort.py [hunk 2]
```diff
@@ -100,9 +113,11 @@ def gpu_kernels(self, node, nodename):
         param_types = [ga.SIZE] * (ndim - 1)  # dims
         for _ in range(self.return_values + self.return_indices):
             param_types.append(ga.GpuArray)  # dst*
+            param_types.extend([ga.SIZE] * ndim)  # offset
```
Comment (khaotik) [inline theano/gpuarray/sort.py]: This line should be `param_types.append(ga.SIZE)`

Comment (khaotik) [inline theano/gpuarray/sort.py]: This line should be `param_types.append(ga.SIZE)`

Comment (nouiz) [reply theano/gpuarray/sort.py]: Thanks, this fix one error I got yesterday. I pushed this fix. Now I have another error CUDA_ERROR_LAUNCH_OUT_OF_RESOURCES. I'll check later today, I'm in meeting most of the day.

Comment (khaotik) [reply theano/gpuarray/sort.py]: `CUDA_ERROR_LAUNCH_OUT_OF_RESOURCES` triggers when the kernel use too much registers in GPU. Computing array+offset on CPU side, therefore reducing kernel argument count might help.

Comment (abergeron) [reply theano/gpuarray/sort.py]: We can't use too many registers since we limit register usage for the specific card that we are using.

However, it might be the case that we try to launch too many threads at once which might cause this.

File: theano/gpuarray/sort.py (hunk 1 above)
Comment (abergeron) [reply theano/gpuarray/sort.py]: I did it this way to make sure it's at the top since it wasn't clear which file was read first.  I have no objection to moving it.

**Comment by notoraptor** [comment]
Updated. It seems most of problems come from `ga_half` handling in kernel codes. My last commit try to resolve compilations issues detected and related to this handling.

All tests pass on my computer (`theano/gpuarray/tests/test_sort.py`), **but** I need to reduce size and k value for large cases. For example, if size is 2049, I got the following error:
```
test_argtopk_1d_72 (theano.gpuarray.tests.test_sort.Test_GpuTopK) ... ERROR 3: cuLaunchKernel(k->k, gs[0], 1, 1, ls[0], 1, 1, shared, ctx->s, args, NULL): CUDA_ERROR_LAUNCH_OUT_OF_RESOURCES: too many resources requested for launch
ERROR

======================================================================
ERROR: test_argtopk_1d_72 (theano.gpuarray.tests.test_sort.Test_GpuTopK)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/notoraptor/Programmes/anaconda3/envs/python2/lib/python2.7/site-packages/parameterized/parameterized.py", line 392, in standalone_func
    return func(*(a + p.args), **p.kwargs)
  File "/mnt/00447B2C447B2396/Users/notoraptor/mila/dev/git/theano/theano/tensor/tests/test_sort.py", line 328, in test_argtopk_1d
    yval = fn(xval)
  File "/mnt/00447B2C447B2396/Users/notoraptor/mila/dev/git/theano/theano/compile/function_module.py", line 917, in __call__
    storage_map=getattr(self.fn, 'storage_map', None))
  File "/mnt/00447B2C447B2396/Users/notoraptor/mila/dev/git/theano/theano/gof/link.py", line 325, in raise_with_op
    reraise(exc_type, exc_value, exc_trace)
  File "/mnt/00447B2C447B2396/Users/notoraptor/mila/dev/git/theano/theano/compile/function_module.py", line 903, in __call__
    self.fn() if output_subset is None else\
RuntimeError: topk: gpu kernel failed to execute
Apply node that caused the error: GpuTopKOp{axis=-1, sorted=False, return_values=False, return_indices=True, idx_dtype='int32'}(GpuFromHost<None>.0, TensorConstant{1337})
Toposort index: 1
Inputs types: [GpuArrayType<None>(float32, vector), TensorType(int16, scalar)]
Inputs shapes: [(2049,), ()]
Inputs strides: [(4,), ()]
Inputs values: ['not shown', array(1337, dtype=int16)]
Outputs clients: [[HostFromGpu(gpuarray)(GpuTopKOp{axis=-1, sorted=False, return_values=False, return_indices=True, idx_dtype='int32'}.0)]]

HINT: Re-running with most Theano optimization disabled could give you a back-trace of when this node was created. This can be done with by setting the Theano flag 'optimizer=fast_compile'. If that does not work, Theano optimizations can be disabled with 'optimizer=None'.
HINT: Use the Theano flag 'exception_verbosity=high' for a debugprint and storage map footprint of this apply node.

----------------------------------------------------------------------
Ran 71 tests in 1.789s

FAILED (errors=1)
```

I don't yet know how to resolve this case. But with smaller sizes, tests should pass.

I keep original sizes in the commit, to check if this problem occurs with jenkins.

@nouiz @abergeron @lamblin

**Comment by abergeron** [comment]
To fix the "large" case (which is actually not that large), would probably require a rework of either the calling code or the kernel itself and will probably take more time than we have to dedicate to this.

So unless you find an easy solution, my advice is to just let the GPU version disabled and pray someone else will fix it.

**Comment by nouiz** [comment]
I think we could merge it if the tests pass on jenkins. It still work in some cases. Tests have a tendency to cover extreme cases. So I would go to merge this PR without disable it if jenkins pass.

**Comment by f0k** [comment]
> So unless you find an easy solution, my advice is to just let the GPU version disabled and pray someone else will fix it.

Not sure what's the best solution either -- is it possible to disable it in a way that allows `optimizer_including=local_gpua_topkop` to re-enable it? This way hardly any user will find out about it, but at least nobody will get bitten by a cryptic error either.

In any case, it might be good to mention in the docstring of `T.topk()` that GPU support is experimental and works for small inputs only, and how to enable or disable it.

**Comment by abergeron** [comment]
Yes, of course.  We can just not register the optimization in the default passes and leave it there for people who want to enable it manually.

**Comment by notoraptor** [comment]
All tests passed! Note that tests use the "large" cases: https://github.com/abergeron/Theano/blob/375fec9aebdccaae3f480d9a5933eeaafcb879d7/theano/tensor/tests/test_sort.py

So, could the error be related to my graphic card ? It's a GTX 1050.

```
r$ nvidia-smi
Tue Nov 28 15:50:11 2017       
+-----------------------------------------------------------------------------+
| NVIDIA-SMI 384.90                 Driver Version: 384.90                    |
|-------------------------------+----------------------+----------------------+
| GPU  Name        Persistence-M| Bus-Id        Disp.A | Volatile Uncorr. ECC |
| Fan  Temp  Perf  Pwr:Usage/Cap|         Memory-Usage | GPU-Util  Compute M. |
|===============================+======================+======================|
|   0  GeForce GTX 1050    Off  | 00000000:01:00.0 Off |                  N/A |
| N/A   66C    P0    N/A /  N/A |    428MiB /  4041MiB |      1%      Default |
+-------------------------------+----------------------+----------------------+
                                                                               
+-----------------------------------------------------------------------------+
| Processes:                                                       GPU Memory |
|  GPU       PID   Type   Process name                             Usage      |
|=============================================================================|
|    0      1073      G   /usr/lib/xorg/Xorg                           366MiB |
|    0      1873      G   /usr/bin/compiz                               61MiB |
+-----------------------------------------------------------------------------+
```

**Comment by f0k** [comment]
> So, could the error be related to my graphic card ?

Possibly. I can flawlessly process matrices of 2048x1337 or 1337x2048 on a Tesla K40 and a Titan Xp. So I guess the optimization can stay in.

**Comment by notoraptor** [comment]
@f0k ok !

Note: I get the `too many resource` error in a 750Ti, too:
```
test_argtopk_nd_32 (theano.gpuarray.tests.test_sort.Test_GpuTopK) ... ERROR 3: cuLaunchKernel(k->k, gs[0], 1, 1, ls[0], 1, 1, shared, ctx->s, args, NULL): CUDA_ERROR_LAUNCH_OUT_OF_RESOURCES: too many resources requested for launch
ERROR

======================================================================
ERROR: test_argtopk_nd_32 (theano.gpuarray.tests.test_sort.Test_GpuTopK)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/u/boccoset/.conda/envs/boccosetEnv/lib/python2.7/site-packages/parameterized/parameterized.py", line 392, in standalone_func
    return func(*(a + p.args), **p.kwargs)
  File "/u/boccoset/mila/dev/git/theano/theano/tensor/tests/test_sort.py", line 409, in test_argtopk_nd
    yval = fn(xval)
  File "/u/boccoset/mila/dev/git/theano/theano/compile/function_module.py", line 917, in __call__
    storage_map=getattr(self.fn, 'storage_map', None))
  File "/u/boccoset/mila/dev/git/theano/theano/gof/link.py", line 325, in raise_with_op
    reraise(exc_type, exc_value, exc_trace)
  File "/u/boccoset/mila/dev/git/theano/theano/compile/function_module.py", line 903, in __call__
    self.fn() if output_subset is None else\
RuntimeError: topk: gpu kernel failed to execute
Apply node that caused the error: GpuTopKOp{axis=-3, sorted=False, return_values=False, return_indices=True, idx_dtype='int32'}(GpuFromHost<None>.0, TensorConstant{-1})
Toposort index: 1
Inputs types: [GpuArrayType<None>(float32, 3D), TensorType(int8, scalar)]
Inputs shapes: [(2017, 5, 3), ()]
Inputs strides: [(60, 12, 4), ()]
Inputs values: ['not shown', array(-1, dtype=int8)]
Outputs clients: [[HostFromGpu(gpuarray)(GpuTopKOp{axis=-3, sorted=False, return_values=False, return_indices=True, idx_dtype='int32'}.0)]]

HINT: Re-running with most Theano optimization disabled could give you a back-trace of when this node was created. This can be done with by setting the Theano flag 'optimizer=fast_compile'. If that does not work, Theano optimizations can be disabled with 'optimizer=None'.
HINT: Use the Theano flag 'exception_verbosity=high' for a debugprint and storage map footprint of this apply node.

----------------------------------------------------------------------
Ran 119 tests in 33.238s

FAILED (errors=1)
```
Still pass with smaller sizes for this test (test_argtopk_nd) in 750 Ti, no need to reduce sizes for the other tests.

I made a small commit to replace a `assert np.allclose` with `utt.assert_allclose` to allow tests to pass on the 750 Ti. @nouiz @abergeron @lamblin

File: theano/tensor/tests/test_sort.py [hunk 3]
```diff
@@ -330,7 +330,7 @@ def test_argtopk_1d(self, size, k, dtype, sorted, idx_dtype):
         goal = np.argsort(xval)[idx].astype(idx_dtype)
 
         # due to uniqueness, we expect indices same
-        assert np.all(xval[np.sort(yval)] == xval[np.sort(goal)])
+        utt.assert_allclose(xval[np.sort(goal)], xval[np.sort(yval)])
```
Comment (nouiz) [inline theano/tensor/tests/test_sort.py]: I don't think this is a good change due to the comment. Do you understand why this change fix the test?

Comment (notoraptor) [reply theano/tensor/tests/test_sort.py]: No, I don't know why it fixes it.

And I just realized that I cannot reproduce the error because `xval` is randomly generated without a fixed seed.

**Comment by notoraptor** [comment]
I am not able to reproduce the error related to `assert np.all`, so I reverted the commit. Quite strange.

**Comment by notoraptor** [comment]
Updated: one size reduced, to make sure tests pass on a GTX 750 Ti for daily buildbot. @nouiz @abergeron @lamblin

**Comment by notoraptor** [comment]
@nouiz All tests pass here.

**Comment by nouiz** [comment]
jenkins test windows mac
